### PR TITLE
arch: boards: Add support for i.MX RT1052 SoC and development board

### PR DESCRIPTION
**Pull request by MaureenHelm**
Adds support for the new i.MX RT1052. This initial port includes only essential drivers (uart and gpio) to enable the hello_world and basic/blinky samples; additional drivers will come later.

### CODE REVIEW

File: arch/arm/soc/nxp_imx/rt/soc.c [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+{
+	/* Disable all unused peripheral clock */
+	CCM->CCGR0 = 0x00C0000FU;
+	CCM->CCGR1 = 0x30000000U;
+	CCM->CCGR2 = 0xFF3F303FU;
+	CCM->CCGR3 = 0xF0000330U;
+	CCM->CCGR4 = 0x0000FF3CU;
+	CCM->CCGR5 = 0xF003330FU;
+	CCM->CCGR6 = 0x00FC0F00U;
+}
```
Comment (galak) [inline arch/arm/soc/nxp_imx/rt/soc.c]: Can these magic numbers be cleaned up?

File: arch/arm/soc/nxp_imx/rt/soc.c [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	RTWDOG->TOVAL = 0xFFFF;
+	RTWDOG->CS = (uint32_t) ((RTWDOG->CS) & ~RTWDOG_CS_EN_MASK)
+		| RTWDOG_CS_UPDATE_MASK;
+
+	/* Disable Systick which might be enabled by bootrom */
+	if (SysTick->CTRL & SysTick_CTRL_ENABLE_Msk) {
+		SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+	}
+
+#if defined(__ICACHE_PRESENT) && __ICACHE_PRESENT
```
Comment (galak) [inline arch/arm/soc/nxp_imx/rt/soc.c]: when would this not be true?

File: arch/arm/soc/nxp_imx/rt/soc.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	SCB_EnableICache();
+#endif
+#if defined(__DCACHE_PRESENT) && __DCACHE_PRESENT
```
Comment (galak) [inline arch/arm/soc/nxp_imx/rt/soc.c]: when would this not be true?

File: arch/arm/soc/nxp_imx/rt/soc.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef _ASMLANGUAGE
+
+#include <fsl_common.h>
+#include <device.h>
+#include <misc/util.h>
+#include <drivers/rand32.h>
```
Comment (galak) [inline arch/arm/soc/nxp_imx/rt/soc.h]: what do we need rand32.h for?

File: dts/arm/nxp/nxp_rt.dtsi [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+		gpio5: gpio@400c0000 {
+			compatible = "nxp,imx-gpio";
+			reg = <0x400c0000 0x4000>;
+			interrupts = <88 0>, <89 0>;
+			label = "GPIO_5";
+		};
+
+		i2c1: i2c@403f0000 {
+			compatible = "nxp,kinetis-lpi2c";
```
Comment (galak) [inline dts/arm/nxp/nxp_rt.dtsi]: do we have a yaml & driver for this?  If not lets remove the node til we do.

File: scripts/support/jlink.sh [hunk 6]
```diff
@@ -30,8 +30,8 @@ do_debug() {
     ${GDB} ${TUI} ${ELF_NAME} \
 	-ex "target remote :${GDB_PORT}" \
 	-ex 'monitor halt' \
-	-ex 'load' \
-	-ex 'monitor reset'
+	-ex 'monitor reset' \
+	-ex 'load'
```
Comment (galak) [inline scripts/support/jlink.sh]: may also need to update scripts/support/runner/jlink.py

**Review by galak — CHANGES_REQUESTED** [review]
Some comments and questions

File: arch/arm/soc/nxp_imx/rt/soc.c (hunk 1 above)
Comment (MaureenHelm) [reply arch/arm/soc/nxp_imx/rt/soc.c]: Yes

File: arch/arm/soc/nxp_imx/rt/soc.c (hunk 2 above)
Comment (MaureenHelm) [reply arch/arm/soc/nxp_imx/rt/soc.c]: Good point, will remove the condition

File: dts/arm/nxp/nxp_rt.dtsi (hunk 5 above)
Comment (MaureenHelm) [reply dts/arm/nxp/nxp_rt.dtsi]: I have one in the works and originally intended to include it in this PR, but it's not working correctly so I pulled it out.

File: scripts/support/jlink.sh (hunk 6 above)
Comment (MaureenHelm) [reply scripts/support/jlink.sh]: Will check

Comment (MaureenHelm) [reply scripts/support/jlink.sh]: Moved change to jlink.py. @mbolivar

**Comment by mbolivar** [comment]
> Moved change to jlink.py. @mbolivar

Seems reasonable. The original order was copied from the shell script, but this looks better. Has it been tested on boards with permanent flash?

**Comment by MaureenHelm** [comment]
> Has it been tested on boards with permanent flash?

Yes, I tested on frdm-k64f

**Comment by mbolivar** [comment]
>> Has it been tested on boards with permanent flash?
> Yes, I tested on frdm-k64f

No objections from me, thanks!

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+NXP MIMXRT1050-EVK
+##################
+
+Overview
+********
+
+The i.MX RT1050 is a new processor family featuring NXP's advanced
+implementation of the ARM Cortex-M7 Core. It provides high CPU performance and
+best real-time response. The i.MX RT1050 provides various memory interfaces,
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: delete "best".  Sentence reads fine without it and avoids the marketing spin.

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+including SDRAM, Raw NAND FLASH, NOR FLASH, SD/eMMC, Quad SPI, HyperBus and a
+wide range of other interfaces for connecting peripherals, such as WLAN,
+Bluetooth™, GPS, displays, and camera sensors. Same as other i.MX processors,
+i.MX RT1050 also has rich audio and video features, including LCD display,
+basic 2D graphics, camera interface, SPDIF and I2S audio interface.
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: add a comma after SPDIF

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+Overview
+********
+
+The i.MX RT1050 is a new processor family featuring NXP's advanced
+implementation of the ARM Cortex-M7 Core. It provides high CPU performance and
+best real-time response. The i.MX RT1050 provides various memory interfaces,
+including SDRAM, Raw NAND FLASH, NOR FLASH, SD/eMMC, Quad SPI, HyperBus and a
+wide range of other interfaces for connecting peripherals, such as WLAN,
+Bluetooth™, GPS, displays, and camera sensors. Same as other i.MX processors,
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: Change "Same as other" to "As with other"

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  - JTAG 20-pin connector
+  - OpenSDA with DAPLink
+- Sensor
+  - FXOS8700CQ 6-axis e-compass
+  - CMOS camera sensor interface
+- Expansion port
+  - Arduino interface
+- CAN bus connector
+
+For more information about the MIMXRT1050 SoC and MIMXRT1050-EVK board:
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: before the final colon, add ", see these references"

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+The Segger J-Link firmware does not support command line flashing, therefore
+the ``make flash`` build target is not supported.
+
+Debugging
+=========
+
+This example uses the :ref:`hello_world` sample with the
+:ref:`nxp_opensda_jlink` tools. Use the ``make debug`` build target to build
+your Zephyr application, invoke the J-Link GDB server, attach a GDB client, and
+program your Zephyr application to flash. It will leave you at a gdb prompt.
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: change "It will leave you" to "The following commands will do all that, and leave you"

**Review by dbkinder — DISMISSED** [review]
A few clarification edits

File: boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+   :width: 720px
+   :align: center
+   :alt: MIMXRT1050-EVK
+
+Hardware
+********
+
+- MIMXRT1052DVL6A MCU (600 MHz, 512 KB TCM)
+- Memory
+  - 256 KB SDRAM
```
Comment (dbkinder) [inline boards/arm/mimxrt1050_evk/doc/mimxrt1050_evk.rst]: Need a blank like before a sub-list or it's picked up as a continuation of the previous line.

**Comment by MaureenHelm** [comment]
Addressed @dbkinder's doc comments, fixed an incorrect link to the board schematic, and rebased to the latest arm branch.

**Comment by MaureenHelm** [comment]
Updated for cmake

**Review by galak — APPROVED** [review]

**Pull request merged**